File: app/schemas/credit_request.py
"""
Pydantic Schemas for Credit Request (Stage 1 Gateway)
=====================================================

Covers: creation, submission, approval, rejection, cancellation,
listing (with filters), and detail responses.
"""
from pydantic import BaseModel, Field
from typing import Optional, List, Any
from datetime import datetime


# ============================================================
# Request Schemas
# ============================================================

class CreditRequestCreate(BaseModel):
    """POST /credit-requests — create a DRAFT credit request."""
    zone_id: int
    department_id: Optional[int] = None
    section_id: Optional[int] = None
    budget_code: str = Field(..., min_length=1)
    amount_requested: float = Field(..., gt=0)
    description: str = Field(..., min_length=1)
    fiscal_year: str = "1403"
    attachments: Optional[Any] = None
    form_data: Optional[dict] = None
    client_request_id: Optional[str] = None


class CreditRequestApprove(BaseModel):
    """POST /credit-requests/{id}/approve"""
    amount_approved: Optional[float] = None  # None => use amount_requested
    comment: Optional[str] = None


class CreditRequestReject(BaseModel):
    """POST /credit-requests/{id}/reject"""
    reason: str = Field(..., min_length=1)


class CreditRequestCancel(BaseModel):
    """POST /credit-requests/{id}/cancel"""
    reason: Optional[str] = None


# ============================================================
# Response Schemas
# ============================================================

class CreditRequestResponse(BaseModel):
    """Full credit request detail."""
    id: int
    credit_request_code: str
    status: str
    zone_id: int
    zone_title: Optional[str] = None
    department_id: Optional[int] = None
    department_title: Optional[str] = None
    section_id: Optional[int] = None
    section_title: Optional[str] = None
    budget_code: str
    amount_requested: float
    amount_approved: Optional[float] = None
    description: str
    fiscal_year: str
    attachments: Optional[Any] = None
    form_data: Optional[dict] = None
    created_by_id: int
    created_by_name: Optional[str] = None
    reviewed_by_id: Optional[int] = None
    reviewed_by_name: Optional[str] = None
    reviewed_at: Optional[str] = None
    rejection_reason: Optional[str] = None
    used_transaction_id: Optional[int] = None
    version: int
    created_at: Optional[str] = None
    updated_at: Optional[str] = None

    class Config:
        from_attributes = True


class CreditRequestListItem(BaseModel):
    """Abbreviated item for list views."""
    id: int
    credit_request_code: str
    status: str
    zone_title: Optional[str] = None
    budget_code: str
    amount_requested: float
    amount_approved: Optional[float] = None
    description: str
    created_by_name: Optional[str] = None
    used_transaction_id: Optional[int] = None
    created_at: Optional[str] = None


class CreditRequestListResponse(BaseModel):
    """GET /credit-requests — paginated list."""
    items: List[CreditRequestListItem]
    total: int
    page: int
    limit: int


class CreditRequestActionResponse(BaseModel):
    """Response for state-transition actions (submit/approve/reject/cancel)."""
    status: str  # "success"
    message: str
    new_status: str
    credit_request_code: str


class CreditRequestLogEntry(BaseModel):
    """Single audit log entry."""
    id: int
    action: str
    previous_status: str
    new_status: str
    actor_name: Optional[str] = None
    comment: Optional[str] = None
    created_at: Optional[str] = None

File: frontend/src/types/creditRequest.ts
/**
 * Stage 1 Gateway: Credit Request types (درخواست تامین اعتبار)
 */

export type CreditRequestStatus = 'DRAFT' | 'SUBMITTED' | 'APPROVED' | 'REJECTED' | 'CANCELLED';

export interface CreditRequest {
  id: number;
  credit_request_code: string;
  status: CreditRequestStatus;
  zone_id: number;
  zone_title?: string;
  department_id?: number;
  department_title?: string;
  section_id?: number;
  section_title?: string;
  budget_code: string;
  amount_requested: number;
  amount_approved?: number;
  description: string;
  fiscal_year: string;
  attachments?: unknown;
  form_data?: Record<string, unknown>;
  created_by_id: number;
  created_by_name?: string;
  reviewed_by_id?: number;
  reviewed_by_name?: string;
  reviewed_at?: string;
  rejection_reason?: string;
  used_transaction_id?: number;
  version: number;
  created_at?: string;
  updated_at?: string;
}

export interface CreditRequestListItem {
  id: number;
  credit_request_code: string;
  status: CreditRequestStatus;
  zone_title?: string;
  budget_code: string;
  amount_requested: number;
  amount_approved?: number;
  description: string;
  created_by_name?: string;
  used_transaction_id?: number;
  created_at?: string;
}

export interface CreditRequestListResponse {
  items: CreditRequestListItem[];
  total: number;
  page: number;
  limit: number;
}

export interface CreditRequestCreateData {
  zone_id: number;
  department_id?: number;
  section_id?: number;
  budget_code: string;
  amount_requested: number;
  description: string;
  fiscal_year?: string;
  attachments?: unknown;
  form_data?: Record<string, unknown>;
  client_request_id?: string;
}

export interface CreditRequestApproveData {
  amount_approved?: number;
  comment?: string;
}

export interface CreditRequestRejectData {
  reason: string;
}

export interface CreditRequestCancelData {
  reason?: string;
}

export interface CreditRequestActionResponse {
  status: string;
  message: string;
  new_status: CreditRequestStatus;
  credit_request_code: string;
}

export interface CreditRequestLogEntry {
  id: number;
  action: string;
  previous_status: string;
  new_status: string;
  actor_name?: string;
  comment?: string;
  created_at?: string;
}

/** Status display config */
export const CR_STATUS_CONFIG: Record<CreditRequestStatus, { label: string; color: string; bgClass: string }> = {
  DRAFT: { label: 'پیش‌نویس', color: '#6b7280', bgClass: 'bg-gray-100 text-gray-700' },
  SUBMITTED: { label: 'ارسال شده', color: '#d97706', bgClass: 'bg-amber-100 text-amber-700' },
  APPROVED: { label: 'تایید شده', color: '#16a34a', bgClass: 'bg-green-100 text-green-700' },
  REJECTED: { label: 'رد شده', color: '#dc2626', bgClass: 'bg-red-100 text-red-700' },
  CANCELLED: { label: 'لغو شده', color: '#9ca3af', bgClass: 'bg-gray-100 text-gray-500' },
};
